icu4c/source/tools/icuexportdata/icuexportdata.cpp

```diff
@@ -96,6 +166,108 @@ void dumpEnumeratedProperty(UProperty uproperty, FILE* f) {
     usrc_writeUCPTrie(f, shortPropName, utrie.getAlias(), UPRV_TARGET_SYNTAX_TOML);
 }
 
+void dumpScriptExtensionsArray(FILE* f) {
+    IcuToolErrorCode status("icuexportdata: dumpScriptExtensionsArray");
+
+    const char* scFullPropName = u_getPropertyName(UCHAR_SCRIPT, U_LONG_PROPERTY_NAME);
+    const char* scxShortPropName = u_getPropertyName(UCHAR_SCRIPT_EXTENSIONS, U_SHORT_PROPERTY_NAME);
+
+    // We want to use 16-bits for our exported trie of sc/scx data because we
+    // need 12 bits to match the 12 bits of data stored for sc/scx in the trie
+    // in the uprops.icu data file.
+    UCPTrieValueWidth scWidth = UCPTRIE_VALUE_BITS_16;
+
+    // Create a mutable UCPTrie builder populated with Script property values data.
+    const UCPMap* scInvMap = u_getIntPropertyMap(UCHAR_SCRIPT, status);
+    handleError(status, scFullPropName);
+    LocalUMutableCPTriePointer builder(umutablecptrie_fromUCPMap(scInvMap, status));
+    handleError(status, scFullPropName);
+
+    // The values for the output scx companion array
+    std::vector< std::set<uint16_t>> outputDedupVec;
+
+    // The sc/scx companion array is an array of arrays (of script codes)
+    fputs("[script_extensions.script_code_array]\n", f);
+    fputs("[\n", f);
+    for(const int32_t cp : scxCodePoints) {
+        // Get the Script_Extensions value (array of Script codes)
+        const int32_t SCX_ARRAY_CAPACITY = 32;
+        UScriptCode scxScripts[SCX_ARRAY_CAPACITY];
+        int32_t numScripts = uscript_getScriptExtensions(cp, scxScripts, SCX_ARRAY_CAPACITY, status);
+        handleError(status, scFullPropName);
+
+        // Convert the scx array into a set
+        std::set<uint16_t> scriptSet;
+        for(int i = 0; i < numScripts; i++) {
+            scriptSet.insert(scxScripts[i]);
+        }
+
+        // Determine the index into the output array for the sub-array containing the
+        // scx value array for this code point will be found. Insert the scx value array
+        // if necessary.
+
+        int findIndex = -1;
+        int outputIndex = 0;
+        for (auto it = outputDedupVec.begin(); it != outputDedupVec.end(); it++) {
+            if (*it == scriptSet) {
+                findIndex = outputIndex;
+                break;
+            }
+            outputIndex++;
+        }
+
+        if (findIndex < 0) {   
+            outputDedupVec.push_back(scriptSet);
+
+            // TODO: maybe should use usrc_writeArrayOfMostlyInvChars ?
```

Comment:
Yeah, please use the functions from writesrc.cpp as much as possible, and if you need something that's not there, you can add it

Reply:
Done.